0000-zmix/README.md

```diff
@@ -0,0 +1,395 @@
+assuming an ordered list of size `message_count` where each element represents some particular (unknown) secret involved in the proof, the statements can reference elements (that is, secrets) of that list by index.
+If different statements reference the same index, they effectively reference the same secret.
+
+To illustrate this, consider a statement type for signatures.
+In general, signatures contain at least
+
+1. a public key under which the signature can be verified, and
+1. an *ordered* list of values that are either hidden (that is, secret) or
+revealed (that is, public).
+
+For example, the library will offer a statement type for Boneh Boyen
+Shacham (BBS) signatures as follows:
+
+```
+pub enum Statement {
+    SignatureBBS {
+        public_key: Vec<u8>,
+        messages: Vec<HiddenOrRevealedValue>,
+    },
+    ...
+}
+
+pub enum HiddenOrRevealedValue {
+    HiddenValueIndex(usize),
+    RevealedValue(Vec<u8>),
+}
+```
+
+While each revealed value in the signature statement simply contains
+the actual value, hidden values are qualified with an _index_.
+In accordance with the list metaphor described above, this index must be smaller or equal to the overall number of secrets as specified in the field `message_count` in the proof specification.
+
+### Referencing the same secret from different statements
+
+The index-based approach allows for referring to the *same* secret in the
+overall zero-knowledge proof from *different* statements. For example,
+another important statement type are commitments, such as the following
+Pederson commitment:
+```
+pub enum Statement {
+    ...
+    PedersenCommitment {
+        message_index: usize,
+        commitment: Vec<u8>,
+        ...
+    },
+```
+Consider a proof specification with a `message_count` of 2 and the
+following two statements:
+1. a `SignatureBBS` over three values where the first is hidden referring
+to index 0, the second is hidden referring to index 1, and the third is the
+revealed value 42, and
+1. a `PedersenCommitment` referring (via `message_index`) to index 0.
+
+The fact that both the signature and the commitment refer to index 0 means
+that they both refer to the same secret in the overall list of secrets (of
+size `message_count`) involved in the zero-knowledge proof.
+
+### Expressing equality of secrets
+
+The case when *different signature statements* reference the *same* index is special: this means that the underlying secret values in the signatures are *equal*.
+
+## Supported statement types
+
+Zmix plans to offer the following proof modules for the following statement types:
+
+- Signatures
+    - Boneh Boyen Shacham
+    - Pointcheval Saunders
+- Pedersen commitments
+- Bulletproof intervals
+- Bulletproof set membership inclusive and exclusive
+- zk-SNARK set memberships
+- Verifiable encryption
+- Linkable indistinguisable tags (see ia.cr/2011/658) to realize scope pseudonyms
+
+## Structures
+
+### Statements
+
+```rust
+pub enum Statement {
```

Comment:
This enum should have variants for equality/inequality of witnesses across statements

Reply:
Better to generalize the relations as 
- inequality with public value
- inequality among witness
- equality with public value (same as disclose)
- equality among witness
- greater than/less than with public value
- greater than/less than among witness